### ISSUE TITLE
Dismissal animation transition gets dropped

### ISSUE THREAD

**Issue by afrederick1** [quickbirdstudios/RxCoordinator#27]
A change was made in commit c26da8c44a38d4c41192af381afc009c57579024 on the transition-rework branch that introduces an issue where the dismissal animation passed into a `present` transition gets lost.

TransitionPerformer+Transition.swift
Before:
```
    func present(_ viewController: UIViewController, with options: TransitionOptions, animation: Animation?, completion: PresentationHandler?) {
        viewController.transitioningDelegate = animation
        rootViewController.present(viewController, animated: options.animated, completion: completion)
    }
```

After:
```
    func present(_ viewController: UIViewController, with options: TransitionOptions, animation: Animation?, completion: PresentationHandler?) {
        let previousTransitionDelegate = viewController.transitioningDelegate
        if let animation = animation {
            viewController.transitioningDelegate = animation
        }
        rootViewController.present(viewController, animated: options.animated) {
            viewController.transitioningDelegate = previousTransitionDelegate
            completion?()
        }
    }
```

the issue in the new version is `viewController.transitioningDelegate = previousTransitionDelegate`, where if `previousTransitionDelegate` is nil then the side effect is that when `dismiss()` is later called there is no `transitioningDelegate` set to handle the custom transition.

One thing to note is that I am presenting and dismissing from a NavigationCoordinator, not sure if that makes a difference. I tried to get a custom modal dismissal animation working in your sample app as well in the NewsCoordinator but am seeing the same thing, here's the change I made to try and reproduce:

```
    override func prepareTransition(for route: NewsRoute) -> NavigationTransition {
        switch route {
        case .news:
            var vc = NewsViewController.instantiateFromNib()
            let service = MockNewsService()
            let viewModel = NewsViewModelImpl(newsService: service, coordinator: anyRouter)
            vc.bind(to: viewModel)
            return .push(vc)
        case .newsDetail(let news):
            var vc = NewsDetailViewController.instantiateFromNib()
            let vm = NewsDetailViewModelImpl(news: news)
            vc.bind(to: vm)
            let animation = Animation(presentationAnimation: StaticTransitionAnimation.flippingPresentation,
                                      dismissalAnimation: StaticTransitionAnimation.flippingPresentation)
            DispatchQueue.main.asyncAfter(deadline: .now() + .seconds(3)) {
                self.trigger(.close)
            }
            return .present(vc, animation: animation)
        case .close:
            return .dismiss()
        }
    }
```

I also tried passing in an Animation to the `.dismiss()` with no luck.

In TransitionPerformer+Transition.swift the `present` function sets `transitioningDelegate` on the VC you are presenting, whereas in the `dismiss` function the `transitioningDelegate` is set on the `rootViewController` which in my case is `UINavigationController` created by the `NavigationCoordinator`. Passing a value in `animation` to `dismiss()` there has no effect, as I believe the actual VC being presented is the one that needs to have `transitioningDelegate` set.

Perhaps I'm going about this incorrectly, let me know what you think!

**Comment by pauljohanneskraft** [comment]
I'm going to look into this and give you more information once I can reproduce the error - Thanks for mentioning!

**Comment by pauljohanneskraft** [comment]
9336418011cd9024c7623139a049fc0de3bc35e2 handles the problems mentioned above.

It fixes the dismissAnimation to actually use the animation (also included in the tests now) - Thanks for catching that one!

On top of that, we are no longer saving the previousTransitioningDelegate and instead only set it, if an animation is set - otherwise it will stay as it was set before. We originally changed this because we noticed that with UIAlertController once we set the transitioningDelegate to nil, it would not present from the middle of the screen but instead have a regular modal transition from the bottom of the screen.
